### ISSUE TITLE
 Node check_orient_and_dims_clust_mask_node failed to run

### ISSUE THREAD

**Issue by soichih** [dPys/PyNets#404]
@dPys I am trying to test PyNets so that I can run it on brainlife.io eventually.

Right now, I am having a problem running it locally on my dev machine and I am wondering if you could guide me in the right direction. 

I've prepared the following testdata

```
$ tree testdata/
testdata/
├── bold.nii.gz
├── dwi.bvals
├── dwi.bvecs
├── dwi.nii.gz
├── mask.nii.gz
├── regressors.tsv
└── t1.nii.gz
```

I then launched PyNets via singularity like this

```
singularity run -e docker://dpys/pynets:latest pynets \
output -p 1 -mod partcorr corr -min_thr 0.05 -max_thr 0.1 -step_thr 0.01 -sm 0 2 4 -hp 0 0.028 0.080 -ct ward -k 100 200 \
-cm /outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz \
-norm 6 \
-anat '"testdata/t1.nii.gz"' \
-func '"testdata/bold.nii.gz"' \
-conf '"testdata/regressors.tsv"' \
-id brainlife \
-work tmp \
-pm 8,30
```

Here is the ouput/error messages.

```

PyNets Version:
0.9.99c


2020-07-24 01:27:22





------------------------------------------------------------------------

Running workflow for single subject:
brainlife

Using whole-brain pipeline...

Using parcels as nodes...

Applying smoothing to node signal at multiple FWHM mm values: 0, 2, 4...

Applying high-pass filter to node signal at multiple Hz values: 0, 0.028, 0.080...

Extracting node signal using a mean strategy...

Iterating graph estimation across multiple connectivity models: partcorr, corr...

Clustering within mask at multiple resolutions...
Cluster atlas: triple_net_ICA_overlap_3_sig_bin_ward_k100
Cluster atlas: triple_net_ICA_overlap_3_sig_bin_ward_k200

Running fmri connectometry only...
BOLD Image: testdata/bold.nii.gz
BOLD Confound Regressors: testdata/regressors.tsv
T1-Weighted Image:
testdata/t1.nii.gz

-------------------------------------------------------------------------


Parsing functional models...
Running Unimodal Workflow...
200724-01:27:26,84 nipype.workflow INFO:
	 Generated workflow graph: tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/graph.png (graph2use=colored, simple_form=True).

Running with {'n_procs': 8, 'memory_gb': 30, 'scheduler': 'mem_thread'}

200724-01:27:31,515 nipype.workflow INFO:
	 Workflow wf_single_sub_brainlife_fmri_20200724_012723 settings: ['check', 'execution', 'logging', 'monitoring']
200724-01:27:32,365 nipype.workflow INFO:
	 Running in parallel.
200724-01:27:32,381 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 3 jobs ready. Free memory (GB): 30.00/30.00, Free processors: 8/8.
200724-01:27:33,381 nipype.workflow INFO:
	 [MultiProc] Running 3 tasks, and 0 jobs ready. Free memory (GB): 27.30/30.00, Free processors: 4/8.
                     Currently running:
                       * wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node
                       * wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node
                       * wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_clust_mask_node
200724-01:27:34,491 nipype.workflow INFO:
	 [Node] Setting-up "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node" in "/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_func_node".
200724-01:27:34,494 nipype.workflow INFO:
	 [Node] Running "check_orient_and_dims_func_node" ("nipype.interfaces.utility.wrappers.Function")
200724-01:27:34,511 nipype.workflow INFO:
	 [Node] Setting-up "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_clust_mask_node" in "/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_clust_mask_node".
200724-01:27:34,518 nipype.workflow INFO:
	 [Node] Running "check_orient_and_dims_clust_mask_node" ("nipype.interfaces.utility.wrappers.Function")
200724-01:27:34,543 nipype.workflow WARNING:
	 Storing result file without outputs
200724-01:27:34,544 nipype.workflow WARNING:
	 [Node] Error on "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node" (/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_func_node)
200724-01:27:34,585 nipype.workflow INFO:
	 [Node] Setting-up "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node" in "/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_anat_node".
200724-01:27:34,588 nipype.workflow INFO:
	 [Node] Running "check_orient_and_dims_anat_node" ("nipype.interfaces.utility.wrappers.Function")
200724-01:27:34,593 nipype.workflow WARNING:
	 Storing result file without outputs
200724-01:27:34,594 nipype.workflow WARNING:
	 [Node] Error on "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_clust_mask_node" (/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_clust_mask_node)
200724-01:27:34,632 nipype.workflow WARNING:
	 Storing result file without outputs
200724-01:27:34,633 nipype.workflow WARNING:
	 [Node] Error on "wf_single_sub_brainlife_fmri_20200724_012723.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node" (/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200724_012723/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_anat_node)
200724-01:27:35,385 nipype.workflow ERROR:
	 Node check_orient_and_dims_clust_mask_node failed to run on host dev1.soichi.us.
200724-01:27:35,387 nipype.workflow ERROR:
	 Saving crash info to tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/crash-20200724-012735-hayashis-check_orient_and_dims_clust_mask_node-60ce31d5-251b-43c8-a2f1-3fd3a3f894f0.txt
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 516, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 635, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 741, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 397, in run
    runtime = self._run_interface(runtime)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/utility/wrappers.py", line 142, in _run_interface
    out = function_handle(**args)
TypeError: check_orient_and_dims() missing 2 required positional arguments: 'outdir' and 'vox_size'

200724-01:27:35,393 nipype.workflow ERROR:
	 Node check_orient_and_dims_anat_node failed to run on host dev1.soichi.us.
200724-01:27:35,393 nipype.workflow ERROR:
	 Saving crash info to tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/crash-20200724-012735-hayashis-check_orient_and_dims_anat_node-6fcac807-3e79-422d-a461-99f773bc72cb.txt
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nibabel/loadsave.py", line 42, in load
    stat_result = os.stat(filename)
FileNotFoundError: [Errno 2] No such file or directory: 'testdata/t1.nii.gz'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 516, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 635, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 741, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 397, in run
    runtime = self._run_interface(runtime)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/utility/wrappers.py", line 142, in _run_interface
    out = function_handle(**args)
  File "<string>", line 28, in check_orient_and_dims
  File "/opt/conda/lib/python3.6/site-packages/nibabel/loadsave.py", line 44, in load
    raise FileNotFoundError("No such file or no access: '%s'" % filename)
FileNotFoundError: No such file or no access: 'testdata/t1.nii.gz'

200724-01:27:35,399 nipype.workflow ERROR:
	 Node check_orient_and_dims_func_node failed to run on host dev1.soichi.us.
200724-01:27:35,399 nipype.workflow ERROR:
	 Saving crash info to tmp/brainlife_20200724_012725_d2a2be11-bbe1-4bed-80ed-973c2acf26ba_wf_single_subject_fmri_brainlife/crash-20200724-012735-hayashis-check_orient_and_dims_func_node-5db57751-0581-44ea-aee9-d1976aa94ba2.txt
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nibabel/loadsave.py", line 42, in load
    stat_result = os.stat(filename)
FileNotFoundError: [Errno 2] No such file or directory: 'testdata/bold.nii.gz'

```

Thank you for your assistance!

**Comment by dPys** [comment]
You are close @soichih ! 

First, be sure you are up-to-date with the latest commit to master (https://github.com/dPys/PyNets/commit/2c61c05d645c5e6e3e1510741561f29e81dd802d) 

Then, also be sure that you either specify the absolute file paths to pynets only (i.e. relative paths are not supported-- see the note about this in the `Required Inputs` section of the usage docs: https://pynets.readthedocs.io/en/latest/usage.html)

If that doesn't fix it, ping me and we can troubleshoot more.
@dPys

**Comment by soichih** [comment]
@dPYS I tried running it with the 1.0.4 tag, but it looks like the container doesn't exist (yet?)

```
singularity run -e docker://dpys/pynets:1.0.4 pynets
FATAL:   Unable to handle docker://dpys/pynets:1.0.4 uri: failed to get SHA of docker://dpys/pynets:1.0.4: Error reading manifest 1.0.4 in docker.io/dpys/pynets: manifest unknown: manifest unknown
```

Is it possible to get it released on dockerhub?

I also updated the input file paths to use `absolute` paths..

```
singularity run -e docker://dpys/pynets:latest pynets output \
    -p $(jq .p config.json) \
    -mod 'partcorr' 'corr' \
    -min_thr $(jq .min_thr config.json) \
    -max_thr $(jq .max_thr config.json) \
    -step_thr $(jq .step_thr config.json) \
    -sm 0 2 4 -hp 0 0.028 0.080 \
    -ct 'ward' -k 100 200 -cm '/outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz' \
    -norm 6 \
    -anat `pwd`/$(jq .t1 config.json) \
    -func `pwd`/$(jq .bold config.json) \
    -conf `pwd`/$(jq .regressors config.json) \
    -id brainlife \
    -work 'tmp' -pm '8,30'
```

I am still seeing the same error message, so maybe I need to run it with v1.0.4?

**Comment by dPys** [comment]
Yeah, you'll definitely need version 1.0+. I'll push to dockerhub by tomorrow afternoon sometime :-)

**Comment by dPys** [comment]
@soichih  , it's pushed!

**Comment by soichih** [comment]
Great!

I've rerun the job and this time I didn't see the error message I was seeing earlier. However, the pipeline seems to be stuck for the last 2,3 hours it looks like it's not progressing.

```
+ singularity run -e docker://dpys/pynets:latest pynets output -p 1 -mod partcorr corr -min_thr 0.05 -max_thr 0.1 -step_thr 0.01 -sm 0 2 4 -hp 0 0.028 0.080 -ct ward -k 100 200 -cm /home/hayashis/git/PyNets/outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz -norm 6 -anat '/home/hayashis/git/PyNets/"testdata/t1.nii.gz"' -func '/home/hayashis/git/PyNets/"testdata/bold.nii.gz"' -conf '/home/hayashis/git/PyNets/"testdata/regressors.tsv"' -id brainlife -work tmp -pm 8,30
INFO:    Converting OCI blobs to SIF format
INFO:    Starting build...
Getting image source signatures
Copying blob cbd31ae33279 skipped: already exists
Copying blob 5f0d2168db0c skipped: already exists
Copying blob 2adb41c9a03e skipped: already exists
Copying blob 06de214f5ad6 skipped: already exists
Copying config fb3aff91c2 done
Writing manifest to image destination
Storing signatures
2020/07/25 21:35:44  info unpack layer: sha256:cbd31ae332794bb723708526045b062b2fe6a08ccc0f143ea7dc18e0ebe46dea
2020/07/25 21:35:45  warn rootless{usr/local/man} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/25 21:35:46  info unpack layer: sha256:5f0d2168db0c35b9f19451fc5bb7cd72bb7e19bfa198b363a6dcf4b2dd135b40
2020/07/25 21:35:46  info unpack layer: sha256:2adb41c9a03e135ba30df8868fe4ad38679fa26497dae5cc9ddc85e54fdf577f
2020/07/25 21:35:53  warn rootless{usr/lib/x86_64-linux-gnu/gstreamer1.0/gstreamer-1.0/gst-ptp-helper} ignoring (usually) harmless EPERM on setxattr "security.capability"
2020/07/25 21:36:04  info unpack layer: sha256:06de214f5ad6a06b3bc8184f77597ecc710b6365c5ea39af718cadcdbcfc375a
2020/07/25 21:36:41  warn rootless{opt/conda/pkgs/ncurses-6.2-he1b5a44_1/lib/terminfo} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
2020/07/25 21:36:43  warn rootless{opt/conda/pkgs/pypy3.6-7.3.1-h3e02ecb_1/site-packages} ignoring (usually) harmless EPERM on setxattr "user.rootlesscontainers"
INFO:    Creating SIF file...


PyNets
Version: 1.0.05
FSL 6.0 detected: FSLDIR=/usr/share/fsl/5.0
2020-07-25 21:37:39




------------------------------------------------------------------------

Running workflow for single subject: brainlife
Population template: MNI152_T1
Using whole-brain pipeline...
Using parcels as nodes...
Applying smoothing to node signal at multiple FWHM mm values:
0, 2, 4
Applying high-pass filter to node signal at multiple Hz values:
0, 0.028, 0.080
Extracting node signal using a mean strategy...
Iterating graph estimation across multiple connectivity models:
partcorr, corr
Clustering within mask at multiple resolutions:
Cluster atlas:
triple_net_ICA_overlap_3_sig_bin_ward_k100
triple_net_ICA_overlap_3_sig_bin_ward_k200

Running fmri connectometry only...
BOLD Image:
 /home/hayashis/git/PyNets/testdata/bold.nii.gz
BOLD Confound Regressors:
 /home/hayashis/git/PyNets/testdata/regressors.tsv
T1-Weighted Image:
 /home/hayashis/git/PyNets/testdata/t1.nii.gz


-------------------------------------------------------------------------


Parsing functional models...
200725-21:37:42,272 nipype.workflow INFO:
	 Generated workflow graph: tmp/brainlife_20200725_213741_e3c723ef-7bdb-4797-b215-54146087a517_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200725_213740/graph.png (graph2use=colored, simple_form=True).
Running with {'n_procs': 8, 'memory_gb': 30, 'scheduler': 'mem_thread'}

200725-21:37:49,167 nipype.workflow INFO:
	 Workflow wf_single_sub_brainlife_fmri_20200725_213740 settings: ['check', 'execution', 'logging', 'monitoring']



```

I've hit Ctrl+C, and it generated the following output.

```
^CProcess Process-2:
Process Process-2:4:
Process Process-2:2:
Process Process-2:8:
Process Process-2:1:
Process Process-2:7:
Process Process-2:5:
Process Process-2:3:
Process Process-2:6:
Traceback (most recent call last):
  File "/opt/conda/bin/pynets", line 33, in <module>
    sys.exit(load_entry_point('pynets==1.0.5', 'console_scripts', 'pynets')())
  File "/opt/conda/lib/python3.6/site-packages/pynets-1.0.5-py3.6.egg/pynets/cli/pynets_run.py", line 3368, in main
Traceback (most recent call last):
Traceback (most recent call last):
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
Traceback (most recent call last):
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
KeyboardInterrupt
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
KeyboardInterrupt
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
KeyboardInterrupt
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
Traceback (most recent call last):
KeyboardInterrupt
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
  File "/opt/conda/lib/python3.6/concurrent/futures/process.py", line 169, in _process_worker
    call_item = call_queue.get(block=True)
KeyboardInterrupt
  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 93, in get
    with self._rlock:
  File "/opt/conda/lib/python3.6/multiprocessing/synchronize.py", line 95, in __enter__
    return self._semlock.__enter__()
KeyboardInterrupt
    p.join()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 124, in join
    res = self._popen.wait(timeout)
  File "/opt/conda/lib/python3.6/multiprocessing/popen_fork.py", line 50, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/opt/conda/lib/python3.6/multiprocessing/popen_forkserver.py", line 61, in poll
    if not wait([self.sentinel], timeout):
  File "/opt/conda/lib/python3.6/multiprocessing/connection.py", line 911, in wait
    ready = selector.select(timeout)
  File "/opt/conda/lib/python3.6/selectors.py", line 376, in select
    fd_event_list = self._poll.poll(timeout)
KeyboardInterrupt
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/opt/conda/lib/python3.6/site-packages/pynets-1.0.5-py3.6.egg/pynets/cli/pynets_run.py", line 3300, in build_workflow
    wf.run(plugin=plugin_type, plugin_args=plugin_args)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/workflows.py", line 622, in run
    execgraph = generate_expanded_graph(deepcopy(flatgraph))
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 1199, in generate_expanded_graph
    return _remove_nonjoin_identity_nodes(graph_in)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 854, in _remove_nonjoin_identity_nodes
    _remove_identity_node(graph, node)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 883, in _remove_identity_node
    _propagate_root_output(graph, node, field, connections)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/utils.py", line 922, in _propagate_root_output
    destnode.set_input(inport, value)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 302, in set_input
    setattr(self.inputs, parameter, deepcopy(val))
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/traits_extension.py", line 330, in validate
    value = super(File, self).validate(objekt, name, value, return_pathlike=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/traits_extension.py", line 135, in validate
    self.error(objekt, name, str(value))
  File "/opt/conda/lib/python3.6/site-packages/traits/base_trait_handler.py", line 75, in error
    object, name, self.full_info(object, name, value), value
traits.trait_errors.TraitError: The 'clust_mask' trait of a _IndividualClusteringInputSpec instance must be a pathlike object or string representing an existing file, but a value of '/home/hayashis/git/PyNets/outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz' <class 'str'> was specified.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 261, in _bootstrap
    util._exit_function()
  File "/opt/conda/lib/python3.6/multiprocessing/util.py", line 319, in _exit_function
    p.join()
  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 124, in join
    res = self._popen.wait(timeout)
  File "/opt/conda/lib/python3.6/multiprocessing/popen_fork.py", line 50, in wait
    return self.poll(os.WNOHANG if timeout == 0.0 else 0)
  File "/opt/conda/lib/python3.6/multiprocessing/popen_forkserver.py", line 61, in poll
    if not wait([self.sentinel], timeout):
  File "/opt/conda/lib/python3.6/multiprocessing/connection.py", line 911, in wait
    ready = selector.select(timeout)
  File "/opt/conda/lib/python3.6/selectors.py", line 376, in select
    fd_event_list = self._poll.poll(timeout)
KeyboardInterrupt
You have new mail in /var/mail/hayashis

```

Should I give it more time?

**Comment by dPys** [comment]
Glad to hear the issue from before is now resolved @soichih !

It appears that you might be providing a clustering mask to the pipeline that does not exist? 
```
"traits.trait_errors.TraitError: The 'clust_mask' trait of a _IndividualClusteringInputSpec instance must be a pathlike object or string representing an existing file, but a value of '/home/hayashis/git/PyNets/outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz' <class 'str'> was specified."
```

what happens when you `ls -la home/hayashis/git/PyNets/outputs/triple_net_ICA_overlap_3_sig_bin.nii.gz`?

**Comment by soichih** [comment]
Oh.. I mistakenly thought -cm was to specify some output file (since the example path uses `/outputs`). Should I just remove it?

I think I am starting this out too complicated. Is this the simplest example?
> https://pynets.readthedocs.io/en/latest/usage.html#manual-execution-using-the-pynets-cli

I tried running something like it and now seeing this error message.

(both bold.nii.gz and t1.nii.gz exist)

```
singularity run -e docker://dpys/pynets:latest pynets -id brainlife -func /home/hayashis/git/PyNets/testdata/bold.nii.gz -anat /home/hayashis/git/PyNets/testdata/t1.nii.gz -a coords_dosenbach_2010 -mod partcorr -spheres -thr 0.20 -work tmp -pm 8,30 output


PyNets
Version: 1.0.05
FSL 6.0 detected: FSLDIR=/usr/share/fsl/5.0
2020-07-26 18:08:00




------------------------------------------------------------------------

Running workflow for single subject: brainlife
Population template: MNI152_T1
Using whole-brain pipeline...
Using node size of 4mm:
Extracting node signal using a mean strategy...
Using connectivity model: partcorr...

Running fmri connectometry only...
BOLD Image:
 /home/hayashis/git/PyNets/testdata/bold.nii.gz
T1-Weighted Image:
 /home/hayashis/git/PyNets/testdata/t1.nii.gz


-------------------------------------------------------------------------


Parsing functional models...

No functional connectometry iterables...
200726-18:08:03,183 nipype.workflow INFO:
	 Generated workflow graph: tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/graph.png (graph2use=colored, simple_form=True).
Running with {'n_procs': 8, 'memory_gb': 30, 'scheduler': 'mem_thread'}

200726-18:08:07,915 nipype.workflow INFO:
	 Workflow wf_single_sub_brainlife_fmri_20200726_180801 settings: ['check', 'execution', 'logging', 'monitoring']
200726-18:08:07,939 nipype.workflow INFO:
	 Running in parallel.
200726-18:08:07,948 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 2 jobs ready. Free memory (GB): 30.00/30.00, Free processors: 8/8.
200726-18:08:09,720 nipype.workflow INFO:
	 [Node] Setting-up "wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node" in "/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_func_node".
200726-18:08:09,723 nipype.workflow INFO:
	 [Node] Running "check_orient_and_dims_func_node" ("nipype.interfaces.utility.wrappers.Function")
200726-18:08:09,840 nipype.workflow INFO:
	 [Node] Setting-up "wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node" in "/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_anat_node".
200726-18:08:09,842 nipype.workflow INFO:
	 [Node] Running "check_orient_and_dims_anat_node" ("nipype.interfaces.utility.wrappers.Function")
200726-18:08:09,949 nipype.workflow INFO:
	 [MultiProc] Running 2 tasks, and 0 jobs ready. Free memory (GB): 27.50/30.00, Free processors: 5/8.
                     Currently running:
                       * wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node
                       * wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node
Reorienting /home/hayashis/git/PyNets/testdata/bold.nii.gz to RAS+...
200726-18:08:11,515 nipype.workflow WARNING:
	 Storing result file without outputs
200726-18:08:11,516 nipype.workflow WARNING:
	 [Node] Error on "wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_func_node" (/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_func_node)
Reorienting /home/hayashis/git/PyNets/testdata/t1.nii.gz to RAS+...
200726-18:08:11,653 nipype.workflow WARNING:
	 Storing result file without outputs
200726-18:08:11,654 nipype.workflow WARNING:
	 [Node] Error on "wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node" (/mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_anat_node)
200726-18:08:11,953 nipype.workflow ERROR:
	 Node check_orient_and_dims_anat_node failed to run on host dev1.soichi.us.
200726-18:08:11,955 nipype.workflow ERROR:
	 Saving crash info to tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/crash-20200726-180811-hayashis-check_orient_and_dims_anat_node-fe4c8415-364f-4264-b1de-fa57722ab32b.txt
Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 516, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 635, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 741, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 397, in run
    runtime = self._run_interface(runtime)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/utility/wrappers.py", line 142, in _run_interface
    out = function_handle(**args)
  File "<string>", line 61, in check_orient_and_dims
  File "/opt/conda/lib/python3.6/site-packages/pynets-1.0.5-py3.6.egg/pynets/registration/reg_utils.py", line 1392, in reorient_img
    normalized.to_filename(out_name)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/filebasedimages.py", line 334, in to_filename
    self.to_file_map()
  File "/opt/conda/lib/python3.6/site-packages/nibabel/analyze.py", line 1033, in to_file_map
    hdrf = hdr_fh.get_prepare_fileobj(mode='wb')
  File "/opt/conda/lib/python3.6/site-packages/nibabel/fileholders.py", line 70, in get_prepare_fileobj
    obj = ImageOpener(self.filename, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/openers.py", line 114, in __init__
    self.fobj = opener(fileish, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/openers.py", line 54, in _gzip_open
    gzip_file = gzip.GzipFile(filename, mode, compresslevel)
  File "/opt/conda/lib/python3.6/gzip.py", line 163, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'output/pynets/brainlife/func/t1_reor-RAS.nii.gz'
```

The directory `output/pynets/brainlife/func` is empty.

Here is the content of the crash log file

```
$ cat tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/crash-20200726-180811-hayashis-check_orient_and_dims_anat_node-fe4c8415-364f-4264-b1de-fa57722ab32b.txt
Node: wf_single_sub_brainlife_fmri_20200726_180801.meta_wf_brainlife.fmri_connectometry_brainlife.check_orient_and_dims_anat_node
Working directory: /mnt/scratch/hayashis/syncthing/git/PyNets/tmp/brainlife_20200726_180802_107dcc75-a84b-4b5b-836b-1cecd4d8155b_wf_single_subject_fmri_brainlife/wf_single_sub_brainlife_fmri_20200726_180801/meta_wf_brainlife/fmri_connectometry_brainlife/check_orient_and_dims_anat_node

Node inputs:

function_str = def check_orient_and_dims(
        infile,
        outdir,
        vox_size,
        bvecs=None,
        overwrite=True):
    """
    An API to reorient any image to RAS+ and resample any image to a given
    voxel resolution.

    Parameters
    ----------
    infile : str
        File path to a Nifti1Image.
    outdir : str
        Path to base derivatives directory.
    vox_size : str
        Voxel size in mm. (e.g. 2mm).
    bvecs : str
        File path to corresponding bvecs file if infile is a dwi.
    overwrite : bool
        Boolean indicating whether to overwrite existing outputs. Default is
        True.

    Returns
    -------
    outfile : str
        File path to the reoriented and/or resample Nifti1Image.
    bvecs : str
        File path to corresponding reoriented bvecs file if outfile is a dwi.

    """
    from pynets.registration.reg_utils import (
        reorient_dwi,
        reorient_img,
        match_target_vox_res,
    )

    img = nib.load(infile)
    vols = img.shape[-1]

    # Check orientation
    if (vols > 1) and (bvecs is not None):
        # dwi case
        # Check orientation
        if ("reor-RAS" not in infile) or (overwrite is True):
            [infile, bvecs] = reorient_dwi(
                infile, bvecs, outdir, overwrite=overwrite)
        # Check dimensions
        if ("res-" not in infile) or (overwrite is True):
            outfile = match_target_vox_res(
                infile, vox_size, outdir, overwrite=overwrite
            )
            print(outfile)
        else:
            outfile = infile
    elif (vols > 1) and (bvecs is None):
        # func case
        # Check orientation
        if ("reor-RAS" not in infile) or (overwrite is True):
            infile = reorient_img(infile, outdir, overwrite=overwrite)
        # Check dimensions
        if ("res-" not in infile) or (overwrite is True):
            outfile = match_target_vox_res(
                infile, vox_size, outdir, overwrite=overwrite
            )
            print(outfile)
        else:
            outfile = infile
    else:
        # t1w case
        # Check orientation
        if ("reor-RAS" not in infile) or (overwrite is True):
            infile = reorient_img(infile, outdir, overwrite=overwrite)
        # Check dimensions
        if ("res-" not in infile) or (overwrite is True):
            outfile = match_target_vox_res(
                infile, vox_size, outdir, overwrite=overwrite
            )
            print(outfile)
        else:
            outfile = infile

    if bvecs is None:
        return outfile
    else:
        return outfile, bvecs

infile = /home/hayashis/git/PyNets/testdata/t1.nii.gz
outdir = output/pynets/brainlife/func
vox_size = 2mm

Traceback (most recent call last):
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/plugins/multiproc.py", line 67, in run_node
    result["result"] = node.run(updatehash=updatehash)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 516, in run
    result = self._run_interface(execute=True)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 635, in _run_interface
    return self._run_command(execute)
  File "/opt/conda/lib/python3.6/site-packages/nipype/pipeline/engine/nodes.py", line 741, in _run_command
    result = self._interface.run(cwd=outdir)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/base/core.py", line 397, in run
    runtime = self._run_interface(runtime)
  File "/opt/conda/lib/python3.6/site-packages/nipype/interfaces/utility/wrappers.py", line 142, in _run_interface
    out = function_handle(**args)
  File "<string>", line 61, in check_orient_and_dims
  File "/opt/conda/lib/python3.6/site-packages/pynets-1.0.5-py3.6.egg/pynets/registration/reg_utils.py", line 1392, in reorient_img
    normalized.to_filename(out_name)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/filebasedimages.py", line 334, in to_filename
    self.to_file_map()
  File "/opt/conda/lib/python3.6/site-packages/nibabel/analyze.py", line 1033, in to_file_map
    hdrf = hdr_fh.get_prepare_fileobj(mode='wb')
  File "/opt/conda/lib/python3.6/site-packages/nibabel/fileholders.py", line 70, in get_prepare_fileobj
    obj = ImageOpener(self.filename, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/openers.py", line 114, in __init__
    self.fobj = opener(fileish, *args, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/nibabel/openers.py", line 54, in _gzip_open
    gzip_file = gzip.GzipFile(filename, mode, compresslevel)
  File "/opt/conda/lib/python3.6/gzip.py", line 163, in __init__
    fileobj = self.myfileobj = builtins.open(filename, mode or 'rb')
FileNotFoundError: [Errno 2] No such file or directory: 'output/pynets/brainlife/func/t1_reor-RAS.nii.gz'

```

**Comment by dPys** [comment]
I think simplifying for the first ru is probably a good call @soichih . Probably also best to just use labels (i.e. as opposed to spheres) for the first test.

What happens if you just run the following?
```
mkdir -p /home/hayashis/pynets_output
singularity run -e docker://dpys/pynets:latest pynets /home/hayashis/pynets_output id brainlife -func /home/hayashis/git/PyNets/testdata/bold.nii.gz -anat /home/hayashis/git/PyNets/testdata/t1.nii.gz -a 'BrainnetomeAtlasFan2016' -mod partcorr -thr 0.20 -work /tmp -pm 8,30 
 ```

**Comment by soichih** [comment]
IT WORKED!

```
...

Single dataframe for subject brainlife

200726-23:25:32,512 nipype.workflow INFO:
	 [Node] Finished "wf_single_sub_brainlife_fmri_20200726_231946.CombineOutputs".
200726-23:25:34,174 nipype.workflow INFO:
	 [Job 16] Completed (wf_single_sub_brainlife_fmri_20200726_231946.CombineOutputs).
200726-23:25:34,176 nipype.workflow INFO:
	 [MultiProc] Running 0 tasks, and 0 jobs ready. Free memory (GB): 30.00/30.00, Free processors: 8/8.
Cleaning up...


------------FINISHED-----------
Subject:  brainlife
Execution Time:  0:05:52.349457
-------------------------------
```

I basically had to make all paths (output, tmp, all the input files) absolute paths.

Here is the `main`

```
#!/bin/bash
#PBS -l nodes=1:ppn=8,walltime=4:00:00
#PBS -l vmem=30gb
#PBS -N pynet

# This file is used to execute PyNets on brainlife.
# brainlife stages this git repo, writes `config.json` and execute this script.
# this script reads the `config.json` and execute pynets container through singularity

# you can run this script(main) without any parameter to test how this App will run outside brainlife
# you will need to copy config.json.brainlife-sample to config.json before running `main` as `main`
# will read all parameters from config.json

set -x
set -e

mkdir -p output tmp
singularity run -e docker://dpys/pynets:latest pynets \
    `pwd`/output \
    -id brainlife \
    -work `pwd`/tmp \
    -func `pwd`/$(jq -r .bold config.json) \
    -anat `pwd`/$(jq -r .t1 config.json) \
    -a 'BrainnetomeAtlasFan2016' \
    -mod partcorr -thr 0.20 \
    -pm 8,30
```

Should I add back other input files now? 

Do you suggest registering separate Apps for PyNets based on different set of input? Or should I make func/anat required but all other files optional input?

**Comment by dPys** [comment]
Yep, absolute paths only.

But yeah, let's try some more if you're willing? It might also help to break-the-ice a bit for the larger user-base in case others might feel more comfortable to pose additional questions/troubleshoot if needed...

I would say, maybe try this one next?
```
singularity run -e docker://dpys/pynets:latest pynets /home/hayashis/pynets_output id brainlife -func /home/hayashis/git/PyNets/testdata/bold.nii.gz -anat /home/hayashis/git/PyNets/testdata/t1.nii.gz -work /tmp -a 'BrainnetomeAtlasFan2016' 'destrieux2009_rois' -mod partcorr csd -min_thr 0.20 -max_thr 0.80 -step_thr 0.10 -plt -n Default Visual -dwi {absolute path to a preprocessed diffusion image 4d} -bval {absolute path to a bval} -bvec {absolute path to a bvec} 
```
...and assuming that example runs successfully, it might also help to address the other critical questions you posed.

**Comment by soichih** [comment]
I've added dwi input and other configuration options. I think the workflow still runs, but it ran out of memory
> =>> PBS: job killed: vmem 32817496064 exceeded limit 32212254720

I am testing it with a pretty small dwi file.
>   Dimensions:        96 x 96 x 63 x 50

How much memory does PyNets usually require? I am giving 30GB
> #PBS -l vmem=30gb

**Comment by dPys** [comment]
Only 4 cores and 8 GB of memory are needed minimally. However, note that if you specify `-pm 8,30` to the CLI you will probably have to tell your job script to use > 30 GB since a few extra MB are needed to spawn the forkserver, perform basic checks outside of the nipype workflow context. What I would recommend first is to try not specifying `-pm 8,30` in your `pynets` command, and if that doesn't help, try specifying less (e.g. `-pm 8,24`)

**Comment by soichih** [comment]
I removed -pm option, but that made the memory usage to go up even more (?)

> =>> PBS: job killed: vmem 44559609856 exceeded limit 32212254720

So I tried `-pm 8,24` but that didn't work either. 

![image](https://user-images.githubusercontent.com/923896/88617057-b4b01600-d063-11ea-89e6-c8ee9b9095c8.png)

I see this issue happening quite often with python Apps. Does -pm limit the resident memory size, or will it apply to vmem usage also? 

The process count also seems to be exceeding a specified number of cores (8).

![image](https://user-images.githubusercontent.com/923896/88617143-e32df100-d063-11ea-9263-5a0078df42eb.png)

**Comment by dPys** [comment]
Hmm, what happens if you leave out the #PBS -l vmem=30gb option altogether, but keep the `-pm 8,30` and add the `-v` debug flag?

The memory in `-pm` refers to the maximum amount of virtual memory to make available to the workflow. If you leave the `-pm` flag out altogether, PyNets will use the psutil library to automatically detect the number of cores and vmem available on the compute node on which it is running. Though it may not apply, one thing to consider is that most users have only run PyNets in shared memory contexts (1 workflow per subject per compute node). What system is the pipeline currently running on and do you hit the error only when using thee docker container?

Also, are you able to send STDERR/STDOUT or the exact error? 

Hopefully this isn't a bug, but if so, thanks for reporting!

**Comment by soichih** [comment]
If I remove #PBS, it will only allocate the minimum default amount of memory, which I believe is 8GB. 

Unlike stamped2@TACC which only runs 1 jobs per node, most of the resources that we run brainlife jobs (IU HPC, Bridges, Comet, OSG, and other XSEDE resources)  will *stack* jobs based on the available resource on the node and resource requested for each job, so we need to make sure that job will not exceed the requested resource limit.

I just added -v option and rerun the test job
> https://brainlife.io/project/5d64733db29ac960ca2e797f/process/5f1dc1c3a5b6439226a04e76#5f1f41dea5b6431c3ea0e2f8
You can see the latest logs under Raw Output at the bottom.

**Comment by dPys** [comment]
> If I remove #PBS, it will only allocate the minimum default amount of memory, which I believe is 8GB.
> 
> Unlike stamped2@TACC which only runs 1 jobs per node, most of the resources that we run brainlife jobs (IU HPC, Bridges, Comet, OSG, and other XSEDE resources) will _stack_ jobs based on the available resource on the node and resource requested for each job, so we need to make sure that job will not exceed the requested resource limit.
> 
> I just added -v option and rerun the test job
> 

That makes sense. If the currently running job does hit another memory spike, I've pushed two more commits today (https://github.com/dPys/PyNets/commit/ac7901c173b91eeafd4a907be49e0e7acbb831d9 and https://github.com/dPys/PyNets/commit/bd5bfd9221b200cb78a578caa5087523a5c79e0d) that might address it.

> > https://brainlife.io/project/5d64733db29ac960ca2e797f/process/5f1dc1c3a5b6439226a04e76#5f1f41dea5b6431c3ea0e2f8
> > You can see the latest logs under Raw Output at the bottom.

This looks terrific @soichih !!

**Comment by dPys** [comment]
Looking at the logs from the finished job that failed, it looks like all we get is the `OSError: [Errno 30] Read-only file system: '/home/hayashis'`, which seems unrelated (but should be addressed):

Other than that, I did come across this rather old thread: https://neurostars.org/t/fmriprep-vmem-exceeding-limit-on-job/1436 and relatedly this still open issue: https://github.com/poldracklab/fmriprep/issues/857 which appear to be addressing a similar problem encoutered in fMRIprep. In other words, this may be a general problem still in nipype. Perhaps @oesteban or @effigies can chime in on whether fmriprep users still encountering max vmem exceeded errors? 

@soichih -- is the 30gb that max you can make available to a single job on brainlife.io? What happens if you raise the max to `#PBS -l vmem=50gb` for example, but still indicate `-pm 8,30` to the workflow? Just in case, I would also try pulling the latest from dockerhub, since I've been pushing some more changes.

**Comment by oesteban** [comment]
> Perhaps @oesteban or @effigies can chime in on whether fmriprep users still encountering max vmem exceeded errors?

For some reason we don't know, these problems seem to be better handled now since we switched to Python 3.7. I suspect that improvements in the memory handling of forking using multiprocessing did help.

Until Pydra changes the way nipype is executed (ping @satra), I think the best way to keep memory low is to:
1. Switch multiprocessing to "forkserver" mode as soon as possible in your application.
2. Create the pool of workers as soon as possible
3. Create the Nipype workflow in an isolated process (this is rather cumbersome, but saves some memory)
4. Postpone memory-hogs to the latest possible (e.g., @soichih's contribution to fMRIPrep to postpone the pandoc execution).

Other than that, we all need Pydra to see the light!

**Comment by dPys** [comment]
Yeah switching PyNets to forkserver mode was an excellent way to go. 

This --> "2. Create the pool of workers as soon as possible" is interesting and I have a sneaking suspicion given the timing of the spike this is a possible suspect. Though it now appears that the latest run on brainlife is going 7 hours and strong @soichih ? Maybe my I/O patches helped? We should still check the run_stats.log when the pipeline finishes if we can -- it should get saved to the base working nipype directory of the raw outputs when using the `-v` flag.

**Comment by satra** [comment]
@oesteban - pydra has seen the light, but is indeed in the twilight in some areas. for example, for this problem we still need to work on bringing in a local resource manager or extend dask to support memory usage by task.

@dPys - i don't know where your graduation path lies, so don't want to put more on the plate, but perhaps starting to convert some pieces would be helpful. to see an example that separates python functions and the workflow the pydra-ml package repo is a good one to take a look at: tasks.py contains python function definitions, while classifier.py contains the converted pydra tasks and workflows. at this point in the twilight zone, this model looks reasonable to us, but we welcome feedback. of course the more standard decorate a function to be a pydra task is still ok, but separating like this allows reuse of the python functions in non-pydra scripts.

**Comment by dPys** [comment]
@satra ! Good to hear that pydra is coming along. I'd be excited to repackage this in it when the time is right. My graduation target is Spring I think and then I'd hopefully have more time to work on the migration. Curious-- has anyone tried connecting a pydra workflow to a nipype 1.0 workflow? I had actually started to pydra branch of pynets already, but most of it remains scratchwork sitting on my Desktop. I know that @djarecka's seen some of it. And yep I did find pydraML to be immensely helpful so thanks for putting that together.  

@soichih -- Looks like the `-v` flag is actually causing the pipeline to hang and then timeout-- totally separate issue coming from the nipype profiler (that will hopefully be addressed with Pydra). If we drop verbosity/debug on the command on brainlife, we should in the clear. I've also gone ahead and retuned the threading presets in PyNets so that the system load is a bit more manageable. If you ever want to do this yourself by the way, the preset live [here](https://github.com/dPys/PyNets/blob/master/pynets/runconfig.yaml)

**Comment by dPys** [comment]
@soichih -- the multiplex connectome run appears top have finished successfully on brainlife.

![adjacency-supra_sub-brainlife_BrainnetomeAtlasFan2016_mplx_Layer-1_sub-brainlife_modality-dwi_rsn-Vis_model-csd_template-MNI152_T1_nodetype-parc_samples-20000streams_tracktype-local_directget-det_minlength-20_Layer-2](https://user-images.githubusercontent.com/16432683/89113260-4210b300-d434-11ea-8151-ce0427034b37.png)

Closing this issue now, but feel free to reopen if anything else comes up :-)
